File: src/core/datetime.js
/**
 * Normalizes raw date strings from ExifTool to ISO format (YYYY-MM-DDTHH:MM:SS) for file renaming.
 * Preserves "wall clock" time by ignoring timezone offsets to match user expectations.
 *
 * Examples:
 * - "2023:05:20 14:30:00"       -> "2023-05-20T14:30:00"
 * - "2023:05:20 14:30:00-05:00" -> "2023-05-20T14:30:00" (offset ignored)
 * - "2023:05:20 14:30:00Z"      -> "2023-05-20T14:30:00" (Z ignored)
 *
 * Uses regex extraction to avoid timezone drift from Date.parse().
 *
 * @param {string} date - Raw date string from ExifTool
 * @returns {string|null} ISO formatted date string or null if invalid
 */
export function normalizeDiskDate(date) {
  if (!date || typeof date !== "string") {
    return null;
  }

  // Regex to capture: YYYY : MM : DD (space) HH : MM : SS
  // Matches delimiters : or - to be flexible with different tag standards.
  const match = date.trim().match(/^(\d{4})[:/-](\d{2})[:/-](\d{2})\s+(\d{2}):(\d{2}):(\d{2})/);

  if (match) {
    const [_, y, m, d, h, min, s] = match;

    // Return standard ISO format: YYYY-MM-DDTHH:MM:SS
    return `${y}-${m}-${d}T${h}:${min}:${s}`;
  }

  return null;
}

/**
 * Parses an ISO date string into its components.
 *
 * @param {string} isoDate - Date in format "2023-01-15T14:30:45"
 * @returns {Object|null} Parsed components or null if invalid
 *   - year: string (4 digits)
 *   - month: number (1-12)
 *   - day: number (1-31)
 *   - hour: number (0-23)
 *   - minute: number (0-59)
 *   - second: number (0-59)
 */
export function parseIsoDate(isoDate) {
  if (!isoDate || typeof isoDate !== "string") {
    return null;
  }

  const date = new Date(isoDate);

  if (isNaN(date.getTime())) {
    return null;
  }

  return {
    year: String(date.getFullYear()),
    month: date.getMonth() + 1,
    day: date.getDate(),
    hour: date.getHours(),
    minute: date.getMinutes(),
    second: date.getSeconds(),
  };
}

/**
 * Converts a Unix epoch timestamp to both ISO and EXIF datetime formats.
 *
 * @param {string|number} timestamp - Unix epoch timestamp in seconds
 * @returns {{ iso: string, exif: string }} Datetime in both formats:
 *   - iso: "2023-05-20T19:30:00" (for renaming, UTC)
 *   - exif: "2023:05:20 19:30:00+00:00" (for metadata writing, UTC explicit)
 */
export function parsePhotoTakenTime(timestamp) {
  // Parse the timestamp (it's in seconds, not milliseconds)
  const timestampSeconds = parseInt(timestamp);

  // Convert to milliseconds for JavaScript Date constructor
  const timestampMilliseconds = timestampSeconds * 1000;

  // Create Date object (always in UTC since Unix timestamps are UTC-based)
  const date = new Date(timestampMilliseconds);

  // Validate that we got a valid date
  if (isNaN(date.getTime())) {
    throw new Error(`Failed to parse photoTakenTime timestamp: ${timestamp}`);
  }

  // Helper function for padding
  const pad2 = (num) => String(num).padStart(2, "0");

  // Get the parts in UTC
  const YYYY = date.getUTCFullYear();

  // getUTCMonth() is 0-indexed
  const MM = pad2(date.getUTCMonth() + 1);
  const DD = pad2(date.getUTCDate());
  const HH = pad2(date.getUTCHours());
  const mm = pad2(date.getUTCMinutes());
  const ss = pad2(date.getUTCSeconds());

  return {
    iso: `${YYYY}-${MM}-${DD}T${HH}:${mm}:${ss}`,
    exif: `${YYYY}:${MM}:${DD} ${HH}:${mm}:${ss}+00:00`,
  };
}

File: src/pixelkasten/core/datetime.py
"""
Date normalization and parsing utilities for the takeout pipeline.

Ported from packages/core/src/core/datetime.js. These functions handle
EXIF-style dates, Unix epoch timestamps, and ISO date strings.

Key behavior: timezone offsets are stripped early to preserve "wall clock"
time in filenames (e.g., 14:30 stays 14:30 regardless of timezone).
"""

import re
from datetime import datetime, timezone

# Regex to capture: YYYY[:/-]MM[:/-]DD <space> HH:MM:SS
# No $ anchor — subseconds (.123) and timezone (+02:00) after seconds
# are naturally ignored by only capturing up to the 6th group.
_DISK_DATE_RE = re.compile(
    r"^(\d{4})[:/-](\d{2})[:/-](\d{2})\s+(\d{2}):(\d{2}):(\d{2})"
)


def normalize_disk_date(date: str) -> str | None:
    """
    Normalize raw ExifTool date strings to ISO format (YYYY-MM-DDTHH:MM:SS).

    Strips timezone offsets to preserve wall-clock time. Strips subsecond
    precision. Accepts flexible date separators (: - /).

    Returns None for non-string, None, empty, or unparseable input.
    """
    if not isinstance(date, str):
        return None

    m = _DISK_DATE_RE.match(date.strip())
    if not m:
        return None

    y, mo, d, h, mi, s = m.groups()
    return f"{y}-{mo}-{d}T{h}:{mi}:{s}"


def parse_iso_date(iso_date: str) -> dict | None:
    """
    Parse an ISO date string into its components.

    Returns a dict with:
        - year: int
        - month: int (1-12)
        - day: int (1-31)
        - hour: int (0-23)
        - minute: int (0-59)
        - second: int (0-59)

    Returns None for non-string, None, empty, or unparseable input.
    """
    if not isinstance(iso_date, str) or not iso_date:
        return None

    try:
        dt = datetime.fromisoformat(iso_date)
    except ValueError:
        return None

    return {
        "year": dt.year,
        "month": dt.month,
        "day": dt.day,
        "hour": dt.hour,
        "minute": dt.minute,
        "second": dt.second,
    }


def parse_photo_taken_time(timestamp: str | int) -> dict:
    """
    Convert a Unix epoch timestamp (seconds) to ISO and EXIF datetime formats.

    Args:
        timestamp: Unix epoch in seconds (string or int).

    Returns:
        {"iso": "YYYY-MM-DDTHH:MM:SS", "exif": "YYYY:MM:DD HH:MM:SS+00:00"}

    Raises:
        ValueError: If the timestamp cannot be parsed.
    """
    try:
        ts = int(timestamp)
    except (ValueError, TypeError):
        raise ValueError(f"Failed to parse photoTakenTime timestamp: {timestamp}")

    dt = datetime.fromtimestamp(ts, tz=timezone.utc)

    # Validate — int() on "" raises ValueError above, but NaN-equivalent
    # strings like "invalid" also raise, so this is mostly defensive.
    if dt.year < 1:
        raise ValueError(f"Failed to parse photoTakenTime timestamp: {timestamp}")

    iso = dt.strftime("%Y-%m-%dT%H:%M:%S")
    exif = dt.strftime("%Y:%m:%d %H:%M:%S") + "+00:00"

    return {"iso": iso, "exif": exif}
